src/target/stm32mp15.c

```diff
@@ -0,0 +1,154 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This file implements STM32MP15 target specific functions for detecting
+ * the device and providing the XML memory map.
+ *
+ * References:
+ * RM0436 - STM32MP157 advanced Arm®-based 32-bit MPUs, Rev. 5
+ *   https://www.st.com/resource/en/reference_manual/rm0436-stm32mp157-advanced-armbased-32bit-mpus-stmicroelectronics.pdf
+ */
+
+#include "general.h"
+#include "target.h"
+#include "target_internal.h"
+#include "cortexm.h"
+
+/* Access from processor address space.
+ * Access via the debug APB is at 0xe0081000 over AP1. */
+#define DBGMCU_IDCODE      0x50081000U
+#define STM32MP15_UID_BASE 0x5c005234U
+
+#define DBGMCU_IDC (DBGMCU_IDCODE + 0U)
+#define DBGMCU_CR  (DBGMCU_IDCODE + 4U)
+#define DBGSLEEP   (1U << 0U)
+#define DBGSTOP    (1U << 1U)
+#define DBGSTBY    (1U << 2U)
+
+#define ID_STM32MP15x        0x5000U /* RM0436 */
+#define ID_STM32MP15x_ERRATA 0x450U
+
+typedef struct stm32mp15_priv {
+	uint32_t dbg_cr;
+} stm32mp15_priv_s;
+
+static bool stm32mp15_uid(target_s *target, int argc, const char **argv);
+static bool stm32mp15_cmd_rev(target_s *target, int argc, const char **argv);
+
+const command_s stm32mp15_cmd_list[] = {
+	{"uid", stm32mp15_uid, "Print unique device ID"},
+	{"revision", stm32mp15_cmd_rev, "Returns the Device ID and Revision"},
+	{NULL, NULL, NULL},
+};
+
+static void stm32mp15_detach(target_s *target)
+{
+	stm32mp15_priv_s *ps = (stm32mp15_priv_s *)target->target_storage;
+	target_mem_write32(target, DBGMCU_CR, ps->dbg_cr);
+	cortexm_detach(target);
+}
+
+bool stm32mp15_probe(target_s *target)
+{
+	if (target->part_id != ID_STM32MP15x && target->part_id != ID_STM32MP15x_ERRATA)
+		return false;
+
+	target->driver = "STM32MP15";
+	target->attach = cortexm_attach;
+	target->detach = stm32mp15_detach;
+	target_add_commands(target, stm32mp15_cmd_list, target->driver);
+
+	/* Save private storage */
+	stm32mp15_priv_s *priv_storage = calloc(1, sizeof(*priv_storage));
+	priv_storage->dbg_cr = target_mem_read32(target, DBGMCU_CR);
+	target->target_storage = priv_storage;
+
+	/* Build the RAM map */
+	/* Figure 4. Memory map RM0436, pg158 */
+	target_add_ram(target, 0x00000000, 0x10000); /* RETRAM,     64kiB */
+	target_add_ram(target, 0x10000000, 0x20000); /* AHB SRAM1, 128kiB */
+	target_add_ram(target, 0x10020000, 0x20000); /* AHB SRAM2, 128kiB */
+	target_add_ram(target, 0x10040000, 0x10000); /* AHB SRAM3,  64kiB */
+	target_add_ram(target, 0x10050000, 0x10000); /* AHB SRAM4,  64kiB */
```

Comment:
These constants should all be suffixed with `U` to make them unsigned.

Reply:
Applied the item. 
Is it okay to declare so many (contigious) regions? SRAM1-2-3 could be merged, but SRAM4 is by default reserved for CA7 OpenSTLinux DMA chaining and should not be clobbered by CM4. It's up to Developer SDK user to reclaim this back from CA7 to CM4, which is what I did for our system. SRAM3 is often used for OpenAMP remote syscall buffers.

Reply:
It's fine to declare them like this, but given it results in one giant memory region and they're all RAM, it's also absolutely fine for you to define them as one large region to save Flash space.

Further to the original review comment, please define these as proper constants at the top of file to name these otherwise magic numbers. We'd suggest:
```c
STM32MP15_RETRAM_BASE 0x00000000U
STM32MP15_RETRAM_SIZE 0x00010000U
STM32MP15_SRAM_BASE 0x10000000U
STM32MP15_RETRAM_SIZE 0x00060000U
```

Reply:
These regions are visible to CA7 cores at different addresses. For now I'm just reminding that there will be different aliases in cortexa counterpart code. Maybe these macros should be suffixed.

Reply:
That's fine, these were only suggested names. Given the Cortex-A point, we'd insert `_CM4` after the chip name to sort that problem out and be consistent with the naming in the rest of this code.

Reply:
Proposed change: insert `_CM4_` for bases (because of different memory maps) but not for sizes. Clang-formatted.
```diff
diff --git a/src/target/stm32mp15.c b/src/target/stm32mp15.c
index 99955cc6..ae118844 100644
--- a/src/target/stm32mp15.c
+++ b/src/target/stm32mp15.c
@@ -33,10 +33,10 @@
 #include "cortexm.h"
 
 /* Memory map constants for STM32MP15x */
-#define STM32MP15_RETRAM_BASE  0x00000000U
-#define STM32MP15_RETRAM_SIZE  0x00001000U /* RETRAM, 64 KiB */
-#define STM32MP15_AHBSRAM_BASE 0x10000000U
-#define STM32MP15_AHBSRAM_SIZE 0x00060000U /* AHB SRAM 1+2+3+4, 128+128+64+64 KiB */
+#define STM32MP15_CM4_RETRAM_BASE  0x00000000U
+#define STM32MP15_RETRAM_SIZE      0x00001000U /* RETRAM, 64 KiB */
+#define STM32MP15_CM4_AHBSRAM_BASE 0x10000000U
+#define STM32MP15_AHBSRAM_SIZE     0x00060000U /* AHB SRAM 1+2+3+4, 128+128+64+64 KiB */
 
 /* Access from processor address space.
  * Access via the debug APB is at 0xe0081000 over AP1. */
@@ -105,8 +105,8 @@ bool stm32mp15_cm4_probe(target_s *target)
 	target->target_storage = priv;
 
 	/* Figure 4. Memory map from §2.5.2 in RM0436 rev 6, pg158 */
-	target_add_ram(target, STM32MP15_RETRAM_BASE, STM32MP15_RETRAM_SIZE);
-	target_add_ram(target, STM32MP15_AHBSRAM_BASE, STM32MP15_AHBSRAM_SIZE);
+	target_add_ram(target, STM32MP15_CM4_RETRAM_BASE, STM32MP15_RETRAM_SIZE);
+	target_add_ram(target, STM32MP15_CM4_AHBSRAM_BASE, STM32MP15_AHBSRAM_SIZE);
 
 	return true;
 }
```
This doesn't feel right.

Reply:
Iff the sizes between both cores are the same, then this is pretty perfect - if they are not, then prefix the sizes too. Looks good otherwise.

Reply:
I mean... It's the same region. See ST devicetrees, https://github.com/STMicroelectronics/linux/blob/8e756f0b4a005e9a0374ab2ffb20df8c3ed4ed63/arch/arm/boot/dts/stm32mp151.dtsi#L2025-L2029 and more importantly https://github.com/STMicroelectronics/linux/blob/8e756f0b4a005e9a0374ab2ffb20df8c3ed4ed63/arch/arm/boot/dts/stm32mp15xx-dkx.dtsi#L25-L70 for aliases and default OpenAMP remoteproc splits. I doubt the debugger needs to know about both aliases, but just for good measure also referencing this https://wiki.st.com/stm32mpu/wiki/STM32MP15_RAM_mapping

Reply:
This item is still outstanding in the diff - the base addresses still need prefixing as you noted due to them being core-dependant.